File: DataStructures/bst.py
'''
Binary search tree implementation in Python. Methods include insertion, deletion, search, level order traversal, inorder traversal, preorder traversal and postorder traversal.
Time complexities are listed below. The space complexity is O(n).
The delete function is still in progress.

My BST 
              50 
           /     \\ 
          30      70 
         /  \\    /  \
       20   40  60   80 

'''

class Node:
	def __init__(self, data):
		self.left = None
		self.right = None
		self.data = data

#Calculate the height of a tree - from the root to the farthest leaf node
#The time complexity is O(h)
def height(node):
	if node is None:
		return 0
	else:
		lheight = height(node.left)
		rheight = height(node.right)

		if lheight > rheight:
			return lheight + 1
		else:
			return rheight + 1

#Find and return the node with the minimum value in the BST
#The time complexity is O(h)
def minValueNode(root):

	current = root

	#Given the nature of a BST the min node value will be at the bottom left of the tree
	while(current.left is not None):
		current = current.left

	return current

#Best case time complexity is O(logn) and worst case is O(n)
def insert(root, node):
	#Base case for if the root is null
	if root is None:
		root = node
	#Start with looking at the right of the root or node if the new node data is larger than the root data,
	#otherwise, look at the left of the root or new node
	else:
		if root.data < node.data:
			if root.right is None:
				root.right = node
			else:
				insert(root.right, node)
		else:
			if root.left is None:
				root.left = node
			else:
				insert(root.left, node)

#Best case time complexity is O(logn) and worst case is O(n)
def search(root,key):
	#Base Cases: the root is null or the key is the root data
	if root is None or root.data == key:
		return root
	#Start search down the right side of the BST
	if root.data < key:
		return search(root.right, key)
	#Start search down the left side of the BST
	return search(root.left,key)

#Print nodes on a specified level (e.g. level 1 is only the root node)
def printGivenLevel(root, level):
	if root is None:
		return
	if level == 1:
		print("%d" %root.data)
	elif level > 1:
		printGivenLevel(root.left, level - 1)
		printGivenLevel(root.right, level - 1)

#Use both the height and printGivenLevel functions to traverse the BST in a Breadth First Search manner. 
#The printGivenLevel funciton is usd recursively for all levels of the BST. 
def printLevelOrder(root):
	h = height(root)
	for i in range(1, h + 1):
		printGivenLevel(root, i)

#Inorder tree traversal - O(n) where n is the total number of nodes
#This is an example of a Depth First Search
def printInOrder(root):
	if root:
		#First recursively call via the left child
		printInOrder(root.left)
		#Print the data of that node
		print(root.data)
		#Then recursively call the right child
		printInOrder(root.right)

#Preorder tree traversal - O(n) where n is the total number of nodes
#This is an example of a Depth First Search
def printPreOrder(root):
	if root:
		#Print the data of that node
		print(root.data)
		#First recursively call via the left child
		printPreOrder(root.left)
		#Then recursively call the right child
		printPreOrder(root.right)

#Postorder tree traversal - O(n) where n is the total number of nodes
#This is an example of a Depth First Search
def printPostOrder(root):
	if root:
		#First recursively call via the left child
		printPostOrder(root.left)
		#Then recursively call the right child
		printPostOrder(root.right)
		#Print the data of that node
		print(root.data)

def main():
	myRoot = Node(50)
	insert(myRoot, Node(30))
	insert(myRoot, Node(20))
	insert(myRoot, Node(40))
	insert(myRoot, Node(70))
	insert(myRoot, Node(60))
	insert(myRoot, Node(80))

	printInOrder(myRoot)
	print("\n")

	printPreOrder(myRoot)
	print("\n")

	printPostOrder(myRoot)
	print("\n")

	print(search(myRoot,70).data)
	print("\n")

	printLevelOrder(myRoot)

	print("Min value node is {}".format(minValueNode(myRoot).data))


if __name__ == "__main__":
	main()

File: DataStructures/bst.cpp
/*
Binary search tree implementation in C++. Methods so far are insertion, search and inorder traversal (depth first search).
Next methods will be deletion, postorder traversal, preorder traversal and level order traversal.
Time complexities are listed below. The space complexity is O(n)

My BST 
              50 
           /     \\ 
          30      70 
         /  \\    /  \\ 
       20   40  60   80 
*/

#include<iostream>
using namespace std;

class BST{

public:	
	int data;
	BST *left, *right;
	//Default constructor
	BST();
	//Constructor with data value parameter
	BST(int);
	//Declaration for the Hieght function
	int Height(BST *);
	//Declaration for the Insert function
	BST* Insert(BST *, int);
	//Declaration for the Search function
	void Search(BST *, int);
	//Declaration for the printGivenLevel function
	void PrintGivenLevel(BST *, int);
	//Declaration for the printGivenLevel function
	void PrintLevelOrder(BST *);
	//Declaration for the Inorder function
	void Inorder(BST *);
	//Declaration for the Preorder function
	void Preorder(BST *);
	//Declaration for the Postorder function
	void Postorder(BST *);
};

//Default constructor
BST::BST(){
	data = 0;
	left = NULL;
	right = NULL;
}

//Constructor with data value parameter
BST::BST(int value){
	data = value;
	left = right = NULL;
}

//Calculate the height of a tree - from the root to the farthest leaf node
int BST::Height(BST *root){
	if(root == NULL)
		return 0;
	else{
		int lheight = Height(root->left);
		int rheight = Height(root->right);

		if(lheight > rheight)
			return(lheight + 1);
		else
			return(rheight + 1);
	}
}

//Best case time complexity is O(logn) and worst case is O(n)
BST* BST::Insert(BST *root, int value){
	if(!root){
		return new BST(value);
	}
	if(value > root->data){
		root->right = Insert(root->right, value);
	}
	else{
		root->left = Insert(root->left, value);
	}
	return root;
}

//Best case time complexity is O(logn) and worst case is O(n)
void BST::Search(BST *root, int key){
	//If root is null, exit
	if(!root)
		return;
	//If the passed in value matches the node data, then search is complete
	if(root->data == key){
		cout << "Found it: " << root->data << endl;
	}
	//Start search down the right side of the BST
	if(root->data < key){
		Search(root->right, key);
	}
	//Start search down the left side of the BST
	else{
		Search(root->left, key);
	}
}

//Print nodes on a specified level (e.g. level 1 is only the root node)
void BST::PrintGivenLevel(BST *root, int level){
	if(root == NULL)
		return;
	if(level == 1)
		cout << root->data << endl;
	else if (level > 1){
		PrintGivenLevel(root->left, level - 1);
		PrintGivenLevel(root->right, level - 1);
	}
}

//Use both the height and printGivenLevel functions to traverse the BST in a Breadth First Search manner. 
//The printGivenLevel funciton is usd recursively for all levels of the BST. 
void BST::PrintLevelOrder(BST *root){
	int height = Height(root);
	for(int i = 1; i <= height; i++){
		PrintGivenLevel(root, i);
	}
}

//Inorder tree traversal - O(n) where n is the total number of nodes
//This is an example of a Depth First Search
void BST::Inorder(BST *root){
	if(root){
		//First recursively call via the left child
		Inorder(root->left);
		//Print the data of that node
		cout << root->data << endl;
		//Then recursively call the right child
		Inorder(root->right);
	}
}

//Preorder tree traversal - O(n) where n is the total number of nodes
//This is an example of a Depth First Search
void BST::Preorder(BST *root){
	if(root){
		//Print the data of that node
		cout << root->data << endl;
		//First recursively call via the left child
		Preorder(root->left);
		//Then recursively call the right child
		Preorder(root->right);
	}
}

//Postorder tree traversal - O(n) where n is the total number of nodes
//This is an example of a Depth First Search
void BST::Postorder(BST *root){
	if(root){
		//First recursively call via the left child
		Postorder(root->left);
		//Then recursively call the right child
		Postorder(root->right);
		//Print the data of that node
		cout << root->data << endl;
	}
}

int main(){
	BST b, *root, s = NULL;
	root = b.Insert(root, 50);

	b.Insert(root, 30);
	b.Insert(root, 20);
	b.Insert(root, 40);
	b.Insert(root, 70);
	b.Insert(root, 60);
	b.Insert(root, 80);

	b.Inorder(root);
	b.Preorder(root);
	b.Postorder(root);

	b.Search(root, 70);

	cout << "The height of the BST is: " << b.Height(root) << endl;

	b.PrintLevelOrder(root);

	return 0;
}